Build failure: pkgsLLVM.kexec-tools

### Steps To Reproduce

Steps to reproduce the behavior:
1. build `pkgsLLVM.kexec-tools`

### Build log

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/mf8zpph7ah0jxlcafxrypaypzw4cf4gv-kexec-tools-2.0.28.tar.xz
source root is kexec-tools-2.0.28
setting SOURCE_DATE_EPOCH to timestamp 1704972668 of file kexec-tools-2.0.28/kexec-tools.spec
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/4pfhyc13jhz84b8973gxbydhx4fn9vxp-ppc64-elfv2.patch
patching file purgatory/arch/ppc64/Makefile
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config/config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config/config.guess
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/sh"
configure flags: --prefix=/nix/store/xv9rlr1b0kc7qa4bdlr2n72ans49kbnk-kexec-tools-x86_64-unknown-linux-gnu-2.0.28 BUILD_CC=cc --build=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu
checking for x86_64-unknown-linux-gnu-gcc... x86_64-unknown-linux-gnu-gcc
checking whether the C compiler works... no
configure: error: in `/build/kexec-tools-2.0.28':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

No maintainers, pinging @felixsinger from `git blame`.

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.0-18-amd64, Debian GNU/Linux, 12 (bookworm), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.21.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

`config.log` has:
```
configure:2939: aarch64-unknown-linux-gnu-gcc -V >&5
aarch64-unknown-linux-gnu-gcc: error: unrecognized command-line option '-V'
aarch64-unknown-linux-gnu-gcc: fatal error: no input files
compilation terminated.
configure:2950: $? = 1
configure:2939: aarch64-unknown-linux-gnu-gcc -qversion >&5
aarch64-unknown-linux-gnu-gcc: error: unrecognized command-line option '-qversion'; did you mean '--version'?
aarch64-unknown-linux-gnu-gcc: fatal error: no input files
compilation terminated.
configure:2950: $? = 1
configure:2939: aarch64-unknown-linux-gnu-gcc -version >&5
aarch64-unknown-linux-gnu-gcc: error: unrecognized command-line option '-version'
aarch64-unknown-linux-gnu-gcc: fatal error: no input files
compilation terminated.
configure:2950: $? = 1
configure:2970: checking whether the C compiler works
configure:2992: aarch64-unknown-linux-gnu-gcc    conftest.c  >&5
/nix/store/v71nlgkf8g6d36gac9r44pmlx54mdql3-binutils-2.41/bin/ld: cannot find crt1.o: No such file or directory
/nix/store/v71nlgkf8g6d36gac9r44pmlx54mdql3-binutils-2.41/bin/ld: cannot find crti.o: No such file or directory
/nix/store/v71nlgkf8g6d36gac9r44pmlx54mdql3-binutils-2.41/bin/ld: cannot find -lgcc_s: No such file or directory
collect2: error: ld returned 1 exit status
configure:2996: $? = 1
configure:3036: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "kexec-tools"
| #define PACKAGE_TARNAME "kexec-tools"
| #define PACKAGE_VERSION "2.0.28"
| #define PACKAGE_STRING "kexec-tools 2.0.28"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
|
| int
| main (void)
| {
|
|   ;
|   return 0;
| }
configure:3041: error: in `/build/kexec-tools-2.0.28':
configure:3043: error: C compiler cannot create executables
See `config.log' for more details
```

need to edit `all-packages.nix` to unpin gccStdenv
```nix
  kexec-tools = callPackage ../os-specific/linux/kexec-tools {
    # clangStdenv fails with
    # purgatory/arch/i386/entry32-16.S:23:2: error: unknown directive
    #  .arch i386
    #  ^
    # purgatory/arch/i386/entry32-16.S:115:11: error: unknown token in expression
    #  ljmp %cs:*(realdest - entry16)
    #           ^
    # make: *** [Makefile:128: purgatory/arch/i386/entry32-16.o] Error 1
    stdenv = gccStdenv;
  };
```

will get past configure but fails at link step for a different reason

--- a/pkgs/os-specific/linux/kexec-tools/default.nix
+++ b/pkgs/os-specific/linux/kexec-tools/default.nix
@@ -23,7 +23,7 @@ stdenv.mkDerivation rec {
       url = "https://github.com/horms/kexec-tools/commit/328de8e00e298f00d7ba6b25dc3950147e9642e6.patch";
       hash = "sha256-wVQI4oV+hBLq3kGIp2+F5J3f6s/TypDu3Xq583KYc3U=";
     })
-  ];
+  ] ++ lib.optional (stdenv.hostPlatform.useLLVM or false) ./fix-purgatory-llvm-libunwind.patch;
 
   hardeningDisable = [ "format" "pic" "relro" "pie" ];
 

--- a/pkgs/os-specific/linux/kexec-tools/fix-purgatory-llvm-libunwind.patch
+++ b/pkgs/os-specific/linux/kexec-tools/fix-purgatory-llvm-libunwind.patch
@@ -1,0 +1,11 @@
+--- a/purgatory/Makefile	2024-06-06 23:59:33.089005832 -0700
++++ b/purgatory/Makefile	2024-06-06 23:59:56.761169892 -0700
+@@ -62,7 +62,7 @@
+ $(PURGATORY): LDFLAGS=$($(ARCH)_PURGATORY_EXTRA_CFLAGS)\
+ 			-Wl,--no-undefined -nostartfiles -nostdlib \
+ 			-nodefaultlibs -e purgatory_start -r \
+-			-Wl,-Map=$(PURGATORY_MAP)
++			-Wl,-Map=$(PURGATORY_MAP) -static
+ 
+ $(PURGATORY): $(PURGATORY_OBJS)
+ 	$(MKDIR) -p $(@D)

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -9474,17 +9474,7 @@ with pkgs;
 
   kestrel = callPackage ../tools/security/kestrel { };
 
-  kexec-tools = callPackage ../os-specific/linux/kexec-tools {
-    # clangStdenv fails with
-    # purgatory/arch/i386/entry32-16.S:23:2: error: unknown directive
-    #  .arch i386
-    #  ^
-    # purgatory/arch/i386/entry32-16.S:115:11: error: unknown token in expression
-    #  ljmp %cs:*(realdest - entry16)
-    #           ^
-    # make: *** [Makefile:128: purgatory/arch/i386/entry32-16.o] Error 1
-    stdenv = gccStdenv;
-  };
+  kexec-tools = callPackage ../os-specific/linux/kexec-tools { };
 
   keepkey-agent = with python3Packages; toPythonApplication keepkey-agent;
 

